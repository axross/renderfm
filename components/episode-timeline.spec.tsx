import { randNumber, randPhrase } from "@ngneat/falso";
import { render, screen, within } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import {
  EpisodeTimeline,
  EpisodeTimelineItem,
  EpisodeTimelineItemProps,
  EpisodeTimelineProps,
} from "~/components/episode-timeline";

describe("<EpisodeTimeline>", () => {
  function setup(additionalProps?: Partial<EpisodeTimelineProps>) {
    render(
      <EpisodeTimeline data-testid="episode-timeline" {...additionalProps}>
        <EpisodeTimelineItem
          label={randPhrase()}
          time={randNumber({ max: 300 })}
          data-testid="item"
        />

        <EpisodeTimelineItem
          label={randPhrase()}
          time={randNumber({ min: 300, max: 600 })}
          data-testid="item"
        />

        <EpisodeTimelineItem
          label={randPhrase()}
          time={randNumber({ min: 600, max: 1000 })}
          data-testid="item"
        />
      </EpisodeTimeline>
    );

    const episodeTimeline = () => screen.getByTestId("episode-timeline");

    return {
      episodeTimeline,
      items: within(episodeTimeline()).getAllByTestId("item"),
    };
  }

  it("takes items as the children and renders them", () => {
    const { episodeTimeline } = setup();

    expect(episodeTimeline()).toMatchSnapshot();
  });
});

describe("<EpisodeTimelineItem>", () => {
  function setup(additionalProps?: Partial<EpisodeTimelineItemProps>) {
    const onTimeClick = jest.fn();

    render(
      <EpisodeTimelineItem
        label={randPhrase()}
        time={139}
        onTimeClick={onTimeClick}
        data-testid="episode-timeline-item"
        {...additionalProps}
      />
    );

    const episodeTimelineItem = () =>
      screen.getByTestId("episode-timeline-item");

    return {
      label: () => within(episodeTimelineItem()).getByTestId("label"),
      time: () => within(episodeTimelineItem()).getByTestId("time"),
      onTimeClick,
    };
  }

  it("renders the label and time", () => {
    const { label, time } = setup();

    expect(label()).toMatchSnapshot();
    expect(time()).toMatchSnapshot();
  });

  it("calls onTimeClick() when the user click the time", () => {
    const { time, onTimeClick } = setup();

    userEvent.click(time());

    expect(onTimeClick).toHaveBeenCalledTimes(1);
    expect(onTimeClick).toHaveBeenNthCalledWith(1, expect.anything(), 139);
  });
});
