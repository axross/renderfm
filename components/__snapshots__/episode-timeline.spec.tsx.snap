// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EpisodeTimeline> takes items as the children and renders them 1`] = `
<ul
  class="flex flex-col gap-y-1 list-disc list-inside"
  data-testid="episode-timeline"
>
  <li
    class="marker:text-slate-400 dark:marker:text-zinc-600"
    data-testid="item"
  >
    <span
      data-testid="label"
    >
      navigating the hard drive wont do anything, we need to synthesize the auxiliary USB circuit!
    </span>
    <span
      class="ml-1"
    >
      [
      <span
        class="outline-none focus:ring-4 focus:ring-yellow-500/50 transition-all rounded-sm text-yellow-600 dark:text-yellow-400 hover:underline cursor-pointer"
        data-testid="time"
        tabindex="0"
      >
        04
        :
        37
      </span>
      ]
    </span>
  </li>
  <li
    class="marker:text-slate-400 dark:marker:text-zinc-600"
    data-testid="item"
  >
    <span
      data-testid="label"
    >
      Use the auxiliary JSON card, then you can copy the optical matrix!
    </span>
    <span
      class="ml-1"
    >
      [
      <span
        class="outline-none focus:ring-4 focus:ring-yellow-500/50 transition-all rounded-sm text-yellow-600 dark:text-yellow-400 hover:underline cursor-pointer"
        data-testid="time"
        tabindex="0"
      >
        08
        :
        41
      </span>
      ]
    </span>
  </li>
  <li
    class="marker:text-slate-400 dark:marker:text-zinc-600"
    data-testid="item"
  >
    <span
      data-testid="label"
    >
      Ill synthesize the primary AI capacitor, that should array the JBOD sensor!
    </span>
    <span
      class="ml-1"
    >
      [
      <span
        class="outline-none focus:ring-4 focus:ring-yellow-500/50 transition-all rounded-sm text-yellow-600 dark:text-yellow-400 hover:underline cursor-pointer"
        data-testid="time"
        tabindex="0"
      >
        16
        :
        11
      </span>
      ]
    </span>
  </li>
</ul>
`;

exports[`<EpisodeTimelineItem> renders the label and time 1`] = `
<span
  data-testid="label"
>
  Use the redundant AGP transmitter, then you can generate the 1080p circuit!
</span>
`;

exports[`<EpisodeTimelineItem> renders the label and time 2`] = `
<span
  class="outline-none focus:ring-4 focus:ring-yellow-500/50 transition-all rounded-sm text-yellow-600 dark:text-yellow-400 hover:underline cursor-pointer"
  data-testid="time"
  tabindex="0"
>
  02
  :
  19
</span>
`;
